plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c

```diff
@@ -422,6 +425,133 @@ fu_synaptics_rmi_v7_device_write_partition(FuSynapticsRmiDevice *self,
 	return TRUE;
 }
 
+GBytes *
+fu_synaptics_rmi_v7_device_get_pubkey(FuSynapticsRmiDevice *self, GError **error)
+{
+	FuSynapticsRmiFlash *flash = fu_synaptics_rmi_device_get_flash(self);
+	FuSynapticsRmiFunction *f34;
+	g_autoptr(GByteArray) req_addr_zero = g_byte_array_new();
+	g_autoptr(GByteArray) req_cmd = g_byte_array_new();
+	g_autoptr(GByteArray) req_partition_id = g_byte_array_new();
+	g_autoptr(GByteArray) req_transfer_length = g_byte_array_new();
+	g_autoptr(GByteArray) res = NULL;
+	g_autoptr(GByteArray) pubkey = g_byte_array_new();
+	gsize key_size = RMI_KEY_SIZE_2K;
+
+	/* f34 */
+	f34 = fu_synaptics_rmi_device_get_function(self, 0x34, error);
+	if (f34 == NULL)
+		return NULL;
+
+	/* set partition id for bootloader 7 */
+	fu_byte_array_append_uint8(req_partition_id, RMI_PARTITION_ID_PUBKEY);
+	if (!fu_synaptics_rmi_device_write(self,
+					   f34->data_base + 0x1,
+					   req_partition_id,
+					   FU_SYNAPTICS_RMI_DEVICE_FLAG_NONE,
+					   error)) {
+		g_prefix_error(error, "failed to write flash partition id: ");
+		return NULL;
+	}
+	fu_byte_array_append_uint16(req_addr_zero, 0x0, G_LITTLE_ENDIAN);
+	if (!fu_synaptics_rmi_device_write(self,
+					   f34->data_base + 0x2,
+					   req_addr_zero,
+					   FU_SYNAPTICS_RMI_DEVICE_FLAG_NONE,
+					   error)) {
+		g_prefix_error(error, "failed to write flash config address: ");
+		return NULL;
+	}
+
+	/* set transfer length */
+	fu_byte_array_append_uint16(req_transfer_length,
+				    key_size / flash->block_size,
+				    G_LITTLE_ENDIAN);
+	if (!fu_synaptics_rmi_device_write(self,
+					   f34->data_base + 0x3,
+					   req_transfer_length,
+					   FU_SYNAPTICS_RMI_DEVICE_FLAG_NONE,
+					   error)) {
+		g_prefix_error(error, "failed to set transfer length: ");
+		return NULL;
+	}
+
+	/* set command to read */
+	fu_byte_array_append_uint8(req_cmd, RMI_FLASH_CMD_READ);
+	if (!fu_synaptics_rmi_device_write(self,
+					   f34->data_base + 0x4,
+					   req_cmd,
+					   FU_SYNAPTICS_RMI_DEVICE_FLAG_NONE,
+					   error)) {
+		g_prefix_error(error, "failed to write command to read: ");
+		return NULL;
+	}
+	if (!fu_synaptics_rmi_device_poll_wait(self, error)) {
+		g_prefix_error(error, "failed to wait: ");
+		return NULL;
+	}
+
+	/* read back entire buffer in blocks */
+	res = fu_synaptics_rmi_device_read(self, f34->data_base + 0x5, (guint32)key_size, error);
+	if (res == NULL) {
+		g_prefix_error(error, "failed to read: ");
+		return NULL;
+	}
+
+	for (guint i = 0; i < res->len; i++)
+		fu_byte_array_append_uint8(pubkey, res->data[res->len - i - 1]);
+
+	/* success */
+	return g_byte_array_free_to_bytes(g_steal_pointer(&pubkey));
+}
+
+gboolean
+fu_synaptics_rmi_v7_device_secure_check(FuSynapticsRmiDevice *self,
+					FuFirmware *firmware,
+					GError **error)
+{
+	g_autoptr(GBytes) pubkey = NULL;
+	g_autoptr(GPtrArray) imgs = NULL;
```

Comment:
normally `g_auto`-stuff comes last in the variable block - by convention at least.